webrender/src/device/gfx/program.rs

```diff
@@ -658,16 +659,18 @@ impl<B: hal::Backend> Program<B> {
         };
         unsafe {
             use std::iter;
-            cmd_buffer.bind_graphics_descriptor_sets(
-                pipeline_layout,
-                if desc_set_per_pass.is_some() { 0 } else { 1 },
-                desc_set_per_pass
-                    .into_iter()
-                    .chain(iter::once(desc_set_per_frame))
-                    .chain(iter::once(desc_set_per_target))
-                    .chain(iter::once(desc_set_per_draw)),
-                &[dynamic_offset],
-            );
+            if rebind_descriptors {
+                cmd_buffer.bind_graphics_descriptor_sets(
```

Comment:
It seems to me that we are still deferring to much work till this function. We need to aim for it to be just `pass.draw()` basically, reducing as much cruft as possible.

For descriptor binding, can't we just do it higher level? I.e. the start of a frame binds its descriptor set and keeps it forever. The start of a target binds its descriptor sets and not touches it. When we break a batch, we can bind the per-draw descriptor set (even unconditionally).

P.S. also, aren't index buffers always the same? we should be just binding it once for the whole frame, ideally, instead of rebinding here

Reply:
> For descriptor binding, can't we just do it higher level? I.e. the start of a frame binds its descriptor set and keeps it forever. The start of a target binds its descriptor sets and not touches it. When we break a batch, we can bind the per-draw descriptor set (even unconditionally).

I think this would be only applicable if we only had one pipeline layout, but currently we have only one. I tried to implement it, but things broke when I try to bind the per frame graphics descriptor sets (for descriptor slot 0), because those are different for each layout, and if I bind three different descriptor sets for the free different layout fir slot 0 only the third one remains active and I get an error for the other two kind of pipeline layouts:
```
UNASSIGNED-CoreValidation-DrawState-DescriptorSetNotBound(ERROR / SPEC): msgNum: 0 - VkPipeline 0x27d uses set #0 but that set is not bound.
    Objects: 1
        [0] 0x22486f180c8, type: 6, name: NULL
```
On the other hand having only one pipeline layout would allow us to make this change also reduces the overhead on descriptor set lookup. The drawback of this would be that we would have to touch the shaders again to move those bindings in a shared shader part. Also I don't know the overhead having those in the shaders, even though they are unused.
I can make this change but I think that would be out of the scope of this PR. I will add this to the list in https://github.com/szeged/webrender/issues/335.

> P.S. also, aren't index buffers always the same? we should be just binding it once for the whole frame, ideally, instead of rebinding here
Those are only related to debug shaders, so those don't have impact on the test benchmarking results. I will create an issue for this to not forget

Reply:
Thank you for the detailed clarification!
When exactly do we have a switch of the pipeline layouts? I recall, for example, that you have a pipeline layout for the clip shader kind. I'd expect them to be together in the alpha pass. Same for the brush ones.
If we do switch the pipeline layout from time to time, we only need to rebind the descriptor sets starting from the first incompatible descriptor set layout. I.e. you can have the same descriptor set layout for slot [0] of all the used pipeline layouts and never need to re-bind it on a pipeline change.
So it still seems hazardly wrong to try rebinding all the sets on each batch.

Reply:
Unfortunately the incompatible descriptor sets are int the first two slots. That's why I mentioned to change the shaders to have a common pipeline layout.

Reply:
Then we'd only change them on the pipeline layout change, which is still not expected to happen often.